OOModel/src/CodeGenerationVisitor.cpp

```diff
@@ -0,0 +1,178 @@
+ **
+ ** Copyright (c) 2011, 2015 ETH Zurich
+ ** All rights reserved.
+ **
+ ** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+ ** following conditions are met:
+ **
+ **    * Redistributions of source code must retain the above copyright notice, this list of conditions and the
+ **      following disclaimer.
+ **    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+ **      following disclaimer in the documentation and/or other materials provided with the distribution.
+ **    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "CodeGenerationVisitor.h"
+
+#include "expressions/ReferenceExpression.h"
+#include "expressions/MetaCallExpression.h"
+
+#include "ModelBase/src/nodes/NameText.h"
+
+#include "declarations/MetaDefinition.h"
+#include "expressions/BooleanLiteral.h"
+
+namespace OOModel {
+
+CodeGenerationVisitor::CodeGenerationVisitor(QMap<QString, Model::Node *> args) : args_{args} {}
+
+void CodeGenerationVisitor::init()
+{
+	addType<ReferenceExpression>(visitReferenceExpression);
+	addType<Model::NameText>(visitNameText);
+	addType<MetaCallExpression>(visitMetaCallExpression);
+}
+
+void CodeGenerationVisitor::visitChildren(Model::Node* n)
+{
+	for (auto child : n->children()) visit(child);
+}
+
+void CodeGenerationVisitor::visitReferenceExpression(CodeGenerationVisitor* v, OOModel::ReferenceExpression* n)
+{
+	auto input = n->name();
+
+	if (!input.contains("##"))
+	{
+		if (auto argument = v->args_[input])
+		{
+			if (auto argumentReference = DCast<ReferenceExpression>(argument))
+			{
+				// case: there exists a replacement and it is another ReferenceExpression
+				// -> copy name (replacing it would cause child nodes of n to disappear)
+
+				n->setName(argumentReference->name());
```

Comment:
There is a discrepancy here:
* In this part of the `if` you don't use `begin/endModification`, you do use them afterwards. This is definitely incorrect.
* In general, if a node is part of a tree that has a manager, using `begin/endModification` is mandatory and your code will fail otherwise. If your code is temporary, (e.g. doesn't have a proper parent yet), then you don't need to use `begin/end`.
* Figure out which case you are dealing with and use the API consistently.